File: Language.Implementation.Patterns/cxx/ch02_basic_parsing_patterns/p02_lexer_2/lexer/Token.cpp
#include "Token.hpp"
#include <iostream>

const std::vector<std::string> Token::tokenNames = {"n/a", "<EOF>", "NAME", "COMMA", "LBRACK", "RBRACK"};

std::string Token::toString() const {
    const std::string& tname = tokenNames[type];
    return "<'" + text + "'," + tname;
}

std::ostream& operator<<(std::ostream& os, const Token& t) {
    os << t.toString();
    return os;
}

File: Language.Implementation.Patterns/python/ch02_basic_parsing_patterns/p02_lexer_2/Token.py

class Token:
    EOF_TYPE = 1
    NAME = 2
    COMMA = 3
    LBRACK = 4
    RBRACK = 5
    tokenNames = ["n/a", "<EOF>", "NAME", "COMMA", "LBRACK", "RBRACK"]

    def __init__(self, type, text):
        self.type = type
        self.text = text

    def __str__(self):
        tname = Token.tokenNames[self.type]
        return "<'" + self.text + "'," + tname + ">"
